source/extensions/filters/http/cache/cache_headers_utils.cc

```diff
@@ -168,6 +168,107 @@ absl::optional<uint64_t> CacheHeadersUtils::readAndRemoveLeadingDigits(absl::str
   return absl::nullopt;
 }
 
+absl::flat_hash_set<std::string> VaryHeader::parseAllowlist() {
+  // TODO(cbdm): Populate the hash_set from
+  // envoy::extensions::filters::http::cache::v3alpha::CacheConfig::allowed_vary_headers.
+  // Need to make sure that the headers we add here are valid values (i.e., not malformed). That
+  // way, we won't have to check this again in isAllowed.
+  return {"accept-encoding"};
+}
+
+bool VaryHeader::isAllowed(const Http::ResponseHeaderMap& headers) {
+  if (!hasVary(headers)) {
+    return true;
+  }
+
+  std::vector<std::string> varied_headers =
+      parseHeaderValue(headers.get(Http::Headers::get().Vary));
+
+  // If the vary value was malformed, it will not be contained in allowed_headers_.
+  for (const std::string& header : varied_headers) {
+    if (!allowed_headers_.contains(header)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool VaryHeader::hasVary(const Http::ResponseHeaderMap& headers) {
+  const Http::HeaderEntry* vary_header = headers.get(Http::Headers::get().Vary);
+  return ((vary_header) && (!vary_header->value().empty()));
+}
+
+std::string VaryHeader::createVaryKey(const Http::HeaderEntry* vary_header,
+                                      const std::vector<const Http::HeaderEntry*>& entry_headers) {
+  if (!vary_header) {
+    return "";
+  }
+
+  ASSERT(vary_header->key() == "vary");
+
+  std::string vary_key = "vary-key:";
+  std::string vary_key_values;
+
+  for (const std::string& header : parseHeaderValue(vary_header)) {
+    absl::StrAppend(&vary_key, header, ";");
```

Comment:
From *my understanding* the header names can't have `;`.
However, there was a previous discussion above (_https://github.com/envoyproxy/envoy/pull/12763#discussion_r474931572_) to use `\n` to separate values between headers, and `\r` to separate values for the same header.

So my next push would use `\r` instead of `;` in this case.

Are you okay with that?

Reply:
`\r` is fine. I agree that using `;` is correct--my only claim is that using `=` (or `\r`) makes the correctness more obvious to a reader with slightly less knowledge/memory of exactly what characters the RFC allows.